hadoop-ozone/ozone-recon/src/main/java/org/apache/hadoop/ozone/recon/tasks/ReconTaskControllerImpl.java

```diff
@@ -0,0 +1,199 @@
+    threadCount = configuration.getInt(OZONE_RECON_TASK_THREAD_COUNT_KEY,
+        OZONE_RECON_TASK_THREAD_COUNT_DEFAULT);
+    executorService = Executors.newFixedThreadPool(threadCount);
+    reconTaskStatusDao = new ReconTaskStatusDao(sqlConfiguration);
+  }
+
+  @Override
+  public void registerTask(ReconDBUpdateTask task) {
+    String taskName = task.getTaskName();
+    LOG.info("Registered task " + taskName + " with controller.");
+
+    // Store task in Task Map.
+    reconDBUpdateTasks.put(taskName, task);
+    // Store Task in Task failure tracker.
+    taskFailureCounter.put(taskName, new AtomicInteger(0));
+    // Create DB record for the task.
+    ReconTaskStatus reconTaskStatusRecord = new ReconTaskStatus(taskName,
+        0L, 0L);
+    reconTaskStatusDao.insert(reconTaskStatusRecord);
+  }
+
+  /**
+   * For every registered task, we try process step twice and then reprocess
+   * once (if process failed twice) to absorb the events. If a task has failed
+   * reprocess call more than 2 times across events, it is unregistered
+   * (blacklisted).
+   * @param events set of events
+   * @throws InterruptedException
+   */
+  @Override
+  public void consumeOMEvents(OMUpdateEventBatch events)
+      throws InterruptedException {
+
+
+    taskSemaphore.acquire();
+
+    try {
+      Collection<Callable<Pair>> tasks = new ArrayList<>();
+      for (Map.Entry<String, ReconDBUpdateTask> taskEntry :
+          reconDBUpdateTasks.entrySet()) {
+        ReconDBUpdateTask task = taskEntry.getValue();
+        tasks.add(() -> task.process(events));
+      }
+
+      List<String> failedTasks = new ArrayList<>();
+      List<Future<Pair>> results = executorService.invokeAll(tasks);
+      for (Future<Pair> f : results) {
+        String taskName = f.get().getLeft().toString();
+        if (!(Boolean)f.get().getRight()) {
+          failedTasks.add(f.get().getLeft().toString());
+        } else {
+          taskFailureCounter.get(taskName).set(0);
+          storeLastCompletedTransaction(taskName, events.getLastEventInfo());
+        }
+
+      }
+
+      //Retry
+      List<String> retryFailedTasks = new ArrayList<>();
+      if (!failedTasks.isEmpty()) {
+        tasks.clear();
+        for (String taskName : failedTasks) {
+          ReconDBUpdateTask task = reconDBUpdateTasks.get(taskName);
+          tasks.add(() -> task.process(events));
+        }
+        results = executorService.invokeAll(tasks);
+        for (Future<Pair> f : results) {
+          String taskName = f.get().getLeft().toString();
+          if (!(Boolean) f.get().getRight()) {
+            LOG.info("Retry step failed for task : " + taskName);
+            retryFailedTasks.add(taskName);
+          } else {
+            taskFailureCounter.get(taskName).set(0);
+            storeLastCompletedTransaction(taskName, events.getLastEventInfo());
+          }
+        }
+      }
+
+      //Reprocess
+      if (!retryFailedTasks.isEmpty()) {
```

Comment:
This is a very heavy operation, add a TODO for queuing up reprocess tasks so they can be executed separately.